Add rop.ret2csu()

CptGibbon opened the pull request:
added a rop.ret2csu() call
should allow you to easily add ret2csu gadgets to an existing ROPchain

CptGibbon committed:
Add rop.ret2csu()

zachriggle commented on pwnlib/rop/rop.py:
```diff
@@ -1206,6 +1206,92 @@ def search(self, move = 0, regs = None, order = 'size'):
+        Examples:
+            >>> rop = ROP(elf64_noPIE)
+            >>> rop.ret2csu(edi, rsi, rdx)
+            >>> print rop.dump()
```
These have to be actual tests, which means that this should print information out.

zachriggle commented on pwnlib/rop/rop.py:
```diff
@@ -1206,6 +1206,92 @@ def search(self, move = 0, regs = None, order = 'size'):
+
+            >>> rop = ROP(elf64_PIE)
+            >>> rop.ret2csu(edi, rsi, rdx, call=elf.got.read)
+            >>> print rop.dump()
```
Same here, this needs to be an actual test.  I'd recommend picking a binary from `pwnlib.data.elf`.

zachriggle commented on pwnlib/rop/rop.py:
```diff
@@ -1206,6 +1206,92 @@ def search(self, move = 0, regs = None, order = 'size'):
+        if self.migrated:
+            log.error('Cannot append to a migrated chain')
+
+        if context.bits != 64:
```
You need to check the `bits` property of the ELF, not `context`.

zachriggle commented on pwnlib/rop/rop.py:
```diff
@@ -1206,6 +1206,92 @@ def search(self, move = 0, regs = None, order = 'size'):
+            log.error('No non-PIE binaries in [elfs], \'call\' parameter is required')
+
+        # 1st gadget: Populate registers for 2nd gadget
+        offset_g1 = 0x5a # delta between __libc_csu_init & 1st gadget
```
Don't hard-code an offset.  It'd be preferable to have a byte array that we search for.

CptGibbon replied:
Gotcha. Went for a capstone approach second time around.

zachriggle commented on pwnlib/rop/rop.py:
```diff
@@ -1206,6 +1206,92 @@ def search(self, move = 0, regs = None, order = 'size'):
+        if not exes:
+            log.error('No non-library binaries in [elfs]')
+
+        for elf in exes:
```
Can we move this logic into the code that finds other gadgets?

I'd prefer if this could be found under `elf.symbols.libc_csu_init`.

CptGibbon replied:
Humour me please I'm thick. Where exactly should this logic be integrated?

zachriggle commented on pwnlib/rop/rop.py:
```diff
@@ -1206,6 +1206,92 @@ def search(self, move = 0, regs = None, order = 'size'):
+        # Resolve __libc_csu_init if candidate binary is stripped
+        if not elf.sym.has_key(u'__libc_csu_init'):
+            if elf.pie:
+                elf.sym[u'__libc_csu_init'] = u64(elf.read(elf.entry + 0x19, 4).ljust(8, '\x00')) + elf.entry + 0x1d
```
Don't use the hard-coded offsets, thanks!

zachriggle commented on pwnlib/rop/rop.py:
```diff
@@ -1206,6 +1206,92 @@ def search(self, move = 0, regs = None, order = 'size'):
+        """Build a ret2csu ROPchain
+
+        Arguments:
+            edi, rsi, rdx: Three primary registers to populate
```
We should probably check that the `edi` value can be packed with `p32()` (i.e., that it's a 32-bit value).

zachriggle commented:
Cool stuff!  Thanks for the contribution.  I've made some comments so that we can get this merged!

zachriggle commented:
Also, as a follow-up, it would be nice if we could integrate this into the logic for `ROP.setRegisters`.  Perhaps in another PR.

Arusekk commented on pwnlib/rop/rop.py:
```diff
@@ -1206,6 +1206,92 @@ def search(self, move = 0, regs = None, order = 'size'):
+
+        for elf in exes:
+            if not elf.pie:
+                if elf.sym.has_key(u'__libc_csu_init'):
```
`x.has_key(y)` should be replaced with `y in x` to allow easier support of python3

CptGibbon replied:
Appreciate it, done.

CptGibbon committed:
Removed magic numbers

Arusekk reviewed (changes requested):
The use of `x.has_key(y)` is obsolete. It is just a silly thing, but could make things easier in the future.

CptGibbon committed:
Replaced
if elf.sym.has_key(u'__libc_csu_init')
with
if u'__libc_csu_init' in elf.symbols

zachriggle commented:
Please merge `dev` or rebase, thanks!

CptGibbon committed:
Merge remote-tracking branch 'upstream/dev' into dev

disconnect3d commented on pwnlib/rop/rop.py:
```diff
@@ -1206,6 +1208,104 @@ def search(self, move = 0, regs = None, order = 'size'):
+                        elf.sym[u'__libc_csu_fini'] = insn.operands[1].imm
+                    if insn.mnemonic == 'mov' and insn.operands[0].reg == X86_REG_RCX:
+                        elf.sym[u'__libc_csu_init'] = insn.operands[1].imm
+                        break
```
Is it okay that we `break` only in this `if` and not in the previous one?

If we would like to break in both ifs we could write it as:

```python
if insn.mnemonic == 'mov' and insn.operands[0].reg in (X86_REG_R8, X86_REG_RCX):
    elf.sym[u'__libc_csu_init'] = insn.operands[1].imm
    break
```

disconnect3d replied:
Same above

CptGibbon replied:
These lines are correct, the arguments to `__libc_start_main()` are populated in this order and we use those instructions to resolve `__libc_csu_fini()` and `__libc_csu_init()`.
Since we know that the address of `__libc_csu_fini()` is resolved first we don't break after finding it, then we break only once `__libc_csu_init()` is resolved.

disconnect3d commented on pwnlib/rop/rop.py:
```diff
@@ -1206,6 +1208,104 @@ def search(self, move = 0, regs = None, order = 'size'):
+                self.raw(insn.address)
+                break
+        self.raw(0x00) # pop rbx
+        self.raw(0x01) # pop rbp
```
if those are eaten by `pop rbx; pop rbp` why don't we fill it? is it overwritten later on or what? (can we mention it in a comment?)

CptGibbon replied:
In ret2csu the instructions:
`add rbx, 0x01`
`cmp rbx, rbp`
must conclude such that the next `jne` instruction does not jump.
I'll add a comment to this effect.

CptGibbon committed:
Merge remote-tracking branch 'upstream/dev' into dev

CptGibbon committed:
Consistently used unicode strings to search for symbols

CptGibbon committed:
Used the Capstone SKIPDATA option to deal with endbr64 instructions

zachriggle commented on pwnlib/rop/rop.py:
```diff
@@ -1299,10 +1300,10 @@ def ret2csu(self, edi, rsi, rdx, rbx=0, rbp=0, r12=0, r13=0, r14=0, r15=0, call=
             log.error('No non-PIE binaries in [elfs], \'call\' parameter is required')
 
-        csu_function = elf.read(elf.sym['__libc_csu_init'], elf.sym['__libc_csu_fini'] - elf.sym['__libc_csu_init'])
+        csu_function = elf.read(elf.sym[u'__libc_csu_init'], elf.sym[u'__libc_csu_fini'] - elf.sym[u'__libc_csu_init'])
```
I feel like there's something broken elsewhere in Pwntools if we need to explicitly call out `unicode` strings in Python2.

Do you have a test case that fails without the `u'...'` prefix?

CptGibbon replied:
So far as I can tell this does not cause any failures.
I started using the `u'...'` prefix because `elf.sym` is a `dotdict` where the keys are of type `unicode` rather than `str`. I used it in some places but not others so purely for consistency I've used it across the function now.

CptGibbon committed:
Added comment to clarify why we set rbp and rbx

CptGibbon committed:
Account for newer versions of gcc populating registers in a different
order

CptGibbon committed:
Merge remote-tracking branch 'upstream/dev' into dev
